# Day 3 - Toboggan Trajectory

With the toboggan login problems resolved, you set off toward the airport. While travel by toboggan might be easy, it's certainly not safe: there's very minimal steering and the area is covered in trees. You'll need to see which angles will take you near the fewest trees.

Due to the local geology, trees in this area only grow on exact integer coordinates in a grid. You make a map (your puzzle input) of the open squares (.) and trees (#) you can see. For example:

..##.......
#...#...#..
.#....#..#.
..#.#...#.#
.#...##..#.
..#.##.....
.#.#.#....#
.#........#
#.##...#...
#...##....#
.#..#...#.#

These aren't the only trees, though; due to something you read about once involving arboreal genetics and biome stability, the same pattern repeats to the right many times:

..##.........##.........##.........##.........##.........##.......  --->
#...#...#..#...#...#..#...#...#..#...#...#..#...#...#..#...#...#..
.#....#..#..#....#..#..#....#..#..#....#..#..#....#..#..#....#..#.
..#.#...#.#..#.#...#.#..#.#...#.#..#.#...#.#..#.#...#.#..#.#...#.#
.#...##..#..#...##..#..#...##..#..#...##..#..#...##..#..#...##..#.
..#.##.......#.##.......#.##.......#.##.......#.##.......#.##.....  --->
.#.#.#....#.#.#.#....#.#.#.#....#.#.#.#....#.#.#.#....#.#.#.#....#
.#........#.#........#.#........#.#........#.#........#.#........#
#.##...#...#.##...#...#.##...#...#.##...#...#.##...#...#.##...#...
#...##....##...##....##...##....##...##....##...##....##...##....#
.#..#...#.#.#..#...#.#.#..#...#.#.#..#...#.#.#..#...#.#.#..#...#.#  --->

**You start on the open square (.) in the top-left corner and need to reach the bottom (below the bottom-most row on your map).**

The toboggan can only follow a few specific slopes (you opted for a cheaper model that prefers rational numbers); start by counting all the trees you would encounter for the slope right 3, down 1:

From your starting position at the top-left, check the position that is right 3 and down 1. Then, check the position that is right 3 and down 1 from there, and so on until you go past the bottom of the map.

The locations you'd check in the above example are marked here with O where there was an open square and X where there was a tree:

..##.........##.........##.........##.........##.........##.......  --->
#..O#...#..#...#...#..#...#...#..#...#...#..#...#...#..#...#...#..
.#....X..#..#....#..#..#....#..#..#....#..#..#....#..#..#....#..#.
..#.#...#O#..#.#...#.#..#.#...#.#..#.#...#.#..#.#...#.#..#.#...#.#
.#...##..#..X...##..#..#...##..#..#...##..#..#...##..#..#...##..#.
..#.##.......#.X#.......#.##.......#.##.......#.##.......#.##.....  --->
.#.#.#....#.#.#.#.O..#.#.#.#....#.#.#.#....#.#.#.#....#.#.#.#....#
.#........#.#........X.#........#.#........#.#........#.#........#
#.##...#...#.##...#...#.X#...#...#.##...#...#.##...#...#.##...#...
#...##....##...##....##...#X....##...##....##...##....##...##....#
.#..#...#.#.#..#...#.#.#..#...X.#.#..#...#.#.#..#...#.#.#..#...#.#  --->

In this example, traversing the map using this slope would cause you to encounter 7 trees.

Starting at the top-left corner of your map and following a slope of right 3 and down 1, how many trees would you encounter?

# Part 1

In [1]:
file = 'data/d3_data.txt'
delim = '\n'

"""Opening a file"""
file = open(file,"r") 
counter = 0
  
"""Reading from file"""
content = file.read() 
colist = content.split(delim) 

"""Iterate through each line of the file"""
for i in colist: 
    if i: 
        counter += 1

In [4]:
# test what the slopes look like
print(colist[0])
print(colist[1])

......#...........#...#........
.#.....#...##.......#.....##...


In [2]:
print("This is the length of a single slope")
slope = len(colist[0])
print(slope)

print("This is the number of lines in the file") 
print(counter) 

print(f"This is how far horizontally we'll be after {counter} downhills")
endpoint = counter*3
print(endpoint)

print("This is how many times the map needs to be duplicated")
mapdupe = round(endpoint/slope,2)
print(mapdupe)

This is the length of a single slope
31
This is the number of lines in the file
323
This is how far horizontally we'll be after 323 downhills
969
This is how many times the map needs to be duplicated
31.26


In [3]:
i = 0
x = 0
ls = []

"""traverse the length of the hill and create a list of x coordinates for each slope of the hill"""
while i < counter-1:
    x += 3
    ls.append(x)
    i+=1
print('original x coordinates')
print(len(ls))
print(ls)

"""use modulus to "loop" the string - in this way we account for the infinite landscape"""
mls = [x % 31 for x in ls]
print('modulated x coordinates')
print(len(mls))
print([item for item in mls])

original x coordinates
322
[3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 45, 48, 51, 54, 57, 60, 63, 66, 69, 72, 75, 78, 81, 84, 87, 90, 93, 96, 99, 102, 105, 108, 111, 114, 117, 120, 123, 126, 129, 132, 135, 138, 141, 144, 147, 150, 153, 156, 159, 162, 165, 168, 171, 174, 177, 180, 183, 186, 189, 192, 195, 198, 201, 204, 207, 210, 213, 216, 219, 222, 225, 228, 231, 234, 237, 240, 243, 246, 249, 252, 255, 258, 261, 264, 267, 270, 273, 276, 279, 282, 285, 288, 291, 294, 297, 300, 303, 306, 309, 312, 315, 318, 321, 324, 327, 330, 333, 336, 339, 342, 345, 348, 351, 354, 357, 360, 363, 366, 369, 372, 375, 378, 381, 384, 387, 390, 393, 396, 399, 402, 405, 408, 411, 414, 417, 420, 423, 426, 429, 432, 435, 438, 441, 444, 447, 450, 453, 456, 459, 462, 465, 468, 471, 474, 477, 480, 483, 486, 489, 492, 495, 498, 501, 504, 507, 510, 513, 516, 519, 522, 525, 528, 531, 534, 537, 540, 543, 546, 549, 552, 555, 558, 561, 564, 567, 570, 573, 576, 579, 582, 585, 588, 591, 594, 597, 600, 603, 606

In [5]:
"""for each slope in the hill, access the element based on the modulated x coordinate"""
y = 1
final = []
for i in mls:
    final.append(colist[y][i])
    y+=1

In [7]:
"""answer to part 1"""
print(len(final))
print([item for item in final])
answer1 = final.count('#')
print(f'number of trees on my map: {answer1}')

322
['.', '#', '.', '#', '#', '#', '#', '#', '.', '#', '#', '#', '#', '#', '#', '#', '#', '#', '#', '.', '#', '#', '#', '#', '#', '#', '.', '#', '#', '#', '#', '#', '#', '#', '.', '#', '#', '#', '#', '#', '#', '#', '#', '.', '#', '#', '#', '#', '#', '.', '.', '#', '#', '#', '#', '#', '.', '.', '#', '#', '#', '.', '#', '#', '#', '#', '#', '#', '#', '#', '#', '#', '#', '.', '.', '#', '#', '#', '#', '#', '#', '#', '#', '#', '.', '.', '.', '#', '#', '#', '#', '#', '#', '#', '#', '.', '.', '#', '#', '.', '#', '#', '.', '#', '#', '#', '#', '#', '#', '.', '#', '#', '#', '#', '#', '#', '#', '#', '#', '#', '.', '.', '#', '#', '#', '#', '#', '.', '#', '#', '#', '#', '.', '#', '#', '#', '#', '#', '#', '#', '#', '#', '#', '#', '#', '#', '.', '#', '.', '#', '#', '#', '#', '.', '#', '#', '#', '.', '#', '#', '.', '#', '#', '#', '#', '#', '#', '#', '.', '#', '#', '#', '.', '#', '.', '.', '#', '#', '#', '#', '#', '#', '#', '#', '#', '.', '#', '#', '#', '#', '#', '#', '#', '#', '.', '#', '#', '.', '#', 

# Part 2

Time to check the rest of the slopes - you need to minimize the probability of a sudden arboreal stop, after all.

Determine the number of trees you would encounter if, for each of the following slopes, you start at the top-left corner and traverse the map all the way to the bottom:

* Right 1, down 1.
* Right 3, down 1. (This is the slope you already checked.)
* Right 5, down 1.
* Right 7, down 1.
* Right 1, down 2.

In the above example, these slopes would find 2, 7, 3, 4, and 2 tree(s) respectively; multiplied together, these produce the answer 336.

What do you get if you multiply together the number of trees encountered on each of the listed slopes?

In [53]:
def coordinate(right,down):
    i = 0
    x = 0
    ls = []

    """traverse the length of the hill and create a list of x coordinates for each slope of the hill"""
    while i < counter-1:
        x += right
        ls.append(x)
        i+=1
    #print('original x coordinates')
    #print(len(ls))
    #print(ls)

    """use modulus to "loop" the string - in this way we account for the infinite landscape"""
    mls = [x % 31 for x in ls]
    #print('modulated x coordinates')
    #print(len(mls))
    #print([item for item in mls])
    
    y = down
    final = []
    for i in mls:
        if y < counter:
            final.append(colist[y][i])
            y+=down
        else:
            break
    
    answer = final.count('#')
    return answer

In [55]:
r1d1 = coordinate(1,1)
print(r1d1)
r3d1 = coordinate(3,1)
print(r3d1)
r5d1 = coordinate(5,1)
print(r5d1)
r7d1 = coordinate(7,1)
print(r7d1)
r1d2 = coordinate(1,2)
print(r1d2)

64
259
65
59
35


In [56]:
answer2 = r1d1*r3d1*r5d1*r7d1*r1d2
print(answer2)

2224913600